src/metabase/automagic_dashboards/rules.clj

```diff
@@ -276,79 +276,73 @@
-                                      u/pprint-to-str)
-                              e)))
-       nil))))
-
-(defn load-rules
-  "Load and validate all rules in dir."
-  ([dir]
-   (with-resources fs
-     (load-rules fs dir)))
-  ([fs dir]
-   (when-let [dir (resource-path fs dir)]
-     (with-open [ds (Files/newDirectoryStream dir)]
-       (->> ds
-            (filter #(str/ends-with? (.toString ^Path %) ".yaml"))
-            (keep (partial load-rule fs))
-            doall)))))
-
-(defn -main
-  "Entry point for lein task `validate-automagic-dashboards`"
-  [& _]
-  (dorun (load-rules "tables"))
-  (dorun (load-rules "metrics"))
-  (dorun (load-rules "fields"))
-  (System/exit 0))
+(defn- load-rule
+  [^Path f]
+  (try
+    (let [entity-type (file->entity-type f)]
+      (-> f
+          .toUri
+          slurp
+          yaml/parse-string
+          (assoc :rule entity-type)
+          (update :applies_to #(or % entity-type))
+          rules-validator))
+    (catch Exception e
+      (log/error (format "Error parsing %s:\n%s"
+                         (.getFileName f)
+                         (or (some-> e
+                                     ex-data
+                                     (select-keys [:error :value])
+                                     u/pprint-to-str)
+                             e)))
+      nil)))
+
+(defn- load-rule-dir
+  ([dir] (load-rule-dir dir [] {}))
+  ([dir path rules]
+   (with-open [ds (Files/newDirectoryStream dir)]
+     (reduce (fn [rules ^Path f]
+               (cond
+                 (Files/isDirectory f (into-array java.nio.file.LinkOption []))
+                 (load-rule-dir f (conj path (str (.getFileName f))) rules)
+
+                 (file->entity-type f)
+                 (assoc-in rules (concat path [(file->entity-type f) ::leaf]) (load-rule f))
+
+                 :else
+                 rules))
+             rules
+             ds))))
+
+(defmacro ^:private with-resource
+  [[identifier path] & body]
+  `(let [[jar# path#] (-> ~path .toString (str/split #"!" 2))]
+     (if path#
+       (with-open [^FileSystem fs# (-> jar#
+                                       java.net.URI/create
+                                       (FileSystems/newFileSystem (java.util.HashMap.)))]
+         (let [~identifier (.getPath fs# path# (into-array String []))]
+           ~@body))
+       (let [~identifier (.getPath (FileSystems/getDefault) (.getPath ~path) (into-array String []))]
+         ~@body))))
+
+(def ^:private rules
+  (with-resource [path (-> rules-dir io/resource .toURI)]
+    (into {} (load-rule-dir path))))
+
+(defn get-rules
+  "Get all rules with prefix `path`.
+   Path needs to match the entire prefix, so [\"table\"] will match table/TransactionTable.yaml,
+   but not table/TransactionTable/ByCountry.yaml"
+  [path]
+  (->> path
+       (get-in rules)
+       (keep (comp ::leaf val))))
+
+(defn get-rule
+  "Get rule at path `path`."
+  [path]
+  (get-in rules (concat path [::leaf])))
```

Comment:
Both of these function have the same signature, but different results. What happens if I pass in a prefix to `get-rule` or a leaf to `get-rules`?

Reply:
(get-rules ["table" "TransactionTable"]) => all indepth rules for TransactionTable
(get-rule ["table" "TransactionTable"]) => TransactionTable rule

Confusing as fuck, agreed. I've renamed one `path` to `prefix`, but probably the names shouldn't be that similar. I'll try to think of better naming.